zkevm-circuits/src/evm_circuit/execution/callop.rs

```diff
@@ -217,13 +239,101 @@ impl<F: Field> ExecutionGadget<F> for CallOpGadget<F> {
             all_but_one_64th_gas,
         );
 
-        // TODO: Handle precompiled
-
         let stack_pointer_delta =
             select::expr(is_call.expr() + is_callcode.expr(), 6.expr(), 5.expr());
         let memory_expansion = call_gadget.memory_expansion.clone();
+
+        // handle precompile calls.
+        let precompile_gadget = cb.condition(
+            and::expr([is_precompile.expr(), is_precheck_ok.expr()]),
+            |cb| {
+                // Write to call's context.
+                for (field_tag, value) in [
+                    (
+                        CallContextFieldTag::IsSuccess,
+                        call_gadget.is_success.expr(),
+                    ),
+                    (
+                        CallContextFieldTag::CalleeAddress,
+                        call_gadget.callee_address_expr(),
+                    ),
+                    (CallContextFieldTag::CallerId, cb.curr.state.call_id.expr()),
+                    (
+                        CallContextFieldTag::CallDataOffset,
+                        call_gadget.cd_address.offset(),
+                    ),
+                    (
+                        CallContextFieldTag::CallDataLength,
+                        call_gadget.cd_address.length(),
+                    ),
+                    (
+                        CallContextFieldTag::ReturnDataOffset,
+                        call_gadget.rd_address.offset(),
+                    ),
+                    (
+                        CallContextFieldTag::ReturnDataLength,
+                        call_gadget.rd_address.length(),
+                    ),
+                ] {
+                    cb.call_context_lookup(
+                        true.expr(),
+                        Some(callee_call_id.expr()),
+                        field_tag,
+                        value,
+                    );
+                }
+
+                // Save caller's call state
+                for (field_tag, value) in [
+                    (CallContextFieldTag::LastCalleeId, callee_call_id.expr()),
+                    (CallContextFieldTag::LastCalleeReturnDataOffset, 0.expr()),
+                    (
+                        CallContextFieldTag::LastCalleeReturnDataLength,
+                        // return_data_len.expr(),
+                        1.expr(), // TODO(rohit): fix this.
+                    ),
+                ] {
+                    cb.call_context_lookup(true.expr(), None, field_tag, value);
+                }
```

Comment:
For each precompile, their return data length should be different, should we set them in the precompile step? Or we can also handle them one by one here.

Reply:
I have updated this part. It was a TODO while I was debugging/fixing something, but later forgot to update the code.